Fix the retrieval of simulator logs

lyndsey-ferguson opened the pull request:
### Description

Fix the retrieval of simulator device logs so that it works on Mac OS 10.11 and no longer fails if the destination directory does not exist or is not normalized.

#### Details

1. Make sure the destination directory exists by using `FileUtils.mkdir_p`
2. Make sure that the destination directory is expanded so that we are not writing to a directory named `~`.
3. Detect if we are running on Mac OS 10.12 before trying to copy a logarchive as they are not useable on Mac OS X 10.11 like they are on Mac OS 10.12. If we are running on 10.12+, copy the logarchive, otherwise copy the system.log like we did before.

### Checklist

- [x] I've run `bundle exec rspec` from the root directory to see all new and existing tests pass
- [ ] I've followed the _fastlane_ code style and run `bundle exec rubocop -a` to ensure the code style is valid
- [x] I've read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary. - N/A

### Description

Added a command to make sure that the logarchive destination directory was fully expanded to handle the case where we get a `~` as the first character and added a command to ensure that the destination directory also exists so that FileUtils.cp_r

### Motivation and Context

Copying the logarchives is failing because the destination directory does not exist. This is not a problem for Ruby 2.2.5, but it is a problem on 2.3.3. Additionally, Mac OS 10.11 does not have the enhanced `logarchive` files that 10.12 has and are no useable.

```
Errno::ENOENT: [!] No such file or directory @ dir_s_mkdir - /Users/jenkins/workspace/workspace/lyndsey-ferguson_MyTests/fastlane/artifacts/tests/system_logs-iPhone 5s_iOS_10.0.logarchive
  /usr/local/Cellar/ruby23/2.3.3/lib/ruby/2.3.0/fileutils.rb:1366:in `mkdir'
  /usr/local/Cellar/ruby23/2.3.3/lib/ruby/2.3.0/fileutils.rb:1366:in `copy'
  /usr/local/Cellar/ruby23/2.3.3/lib/ruby/2.3.0/fileutils.rb:472:in `block in copy_entry'
  /usr/local/Cellar/ruby23/2.3.3/lib/ruby/2.3.0/fileutils.rb:1498:in `wrap_traverse'
  /usr/local/Cellar/ruby23/2.3.3/lib/ruby/2.3.0/fileutils.rb:469:in `copy_entry'
  /usr/local/Cellar/ruby23/2.3.3/lib/ruby/2.3.0/fileutils.rb:444:in `block in cp_r'
  /usr/local/Cellar/ruby23/2.3.3/lib/ruby/2.3.0/fileutils.rb:1571:in `block in fu_each_src_dest'
  /usr/local/Cellar/ruby23/2.3.3/lib/ruby/2.3.0/fileutils.rb:1587:in `fu_each_src_dest0'
  /usr/local/Cellar/ruby23/2.3.3/lib/ruby/2.3.0/fileutils.rb:1569:in `fu_each_src_dest'
  /usr/local/Cellar/ruby23/2.3.3/lib/ruby/2.3.0/fileutils.rb:443:in `cp_r'
  /Users/appian/.gems/gems/fastlane-2.14.2/fastlane_core/lib/fastlane_core/device_manager.rb:208:in `copy_logarchive'
  /Users/appian/.gems/gems/fastlane-2.14.2/scan/lib/scan/runner.rb:101:in `block in copy_simulator_logs'
```

Commit:
Fix retrieval of sim logs.

This addresses 3 issues:
1. The destination directory was not being created.
2. Paths starting with ‘~’ and the FileUtils.cp_r would
fail on Ruby 2.3.3 if the destination directory did not exist.
3. log archives were updated to be viewable for Mac OS 10.12
but are empty for 10.11. Use the old method of copying
the system.log for 10.11 and prior versions

lyndsey-ferguson commented:
@hjanuschka can I get a review please?

hjanuschka reviewed (changes requested):
logarchive has wrong structure now. can you try using `mv`  instead of `mkdir`/`cp_r`

lyndsey-ferguson commented:
@hjanuschka can you look at this approach?

lyndsey-ferguson commented on scan/lib/scan/runner.rb:
```diff
@@ -98,7 +98,7 @@ def copy_simulator_logs
       UI.header("Collecting system logs")
       Scan.devices.each do |device|
         logarchive_dest = File.join(Scan.config[:output_directory], "system_logs-#{device.name}_#{device.os_type}_#{device.os_version}.logarchive")
```
I think I'll change this to just a folder name that has the device details (without system_logs-...etc)

hjanuschka replied:
yeah - like in snapshot, really like the idea.

lyndsey-ferguson commented on snapshot/lib/snapshot/runner.rb:
```diff
@@ -108,7 +108,7 @@ def copy_simulator_logs(device_name, language, locale, launch_arguments)
 
       UI.header("Collecting system logs #{device_name} - #{language}")
       logarchive_dest = File.join(language_folder, "system_logs-" + Digest::MD5.hexdigest(components.join("-")) + ".logarchive")
```
same rename plan

hjanuschka replied:
this one already has the device specific information in the folder name.

lyndsey-ferguson replied:
Well, my plan is to make a _real_ logarchive that is just the name as you have here. Not make an additional folder to hold just `system_logs.logarchive`

lyndsey-ferguson commented on fastlane_core/lib/fastlane_core/device_manager.rb:
```diff
@@ -205,6 +226,7 @@ def copy_logarchive(device, dest)
         command = "xcrun simctl spawn #{device.udid} log collect 2>/dev/null"
         FastlaneCore::CommandExecutor.execute(command: command, print_all: false, print_command: true)
 
+        FileUtils.mkdir_p(dest)
```
This is still necessary on a macOS 10.12 machine

hjanuschka replied:
this causes troubles, on all machines i am able to test this on.

results in a `system_logs-iPhone 6s_iOS_10.2.logarchive` which is a REAL folder - and if open it with console app its empty, if i do `Show package contents`  inside the `system_logs-iPhone 6s_iOS_10.2.logarchive` there is a `system_logs.logarchive`

lyndsey-ferguson replied:
I will figure this out

lyndsey-ferguson commented:
@hjanuschka can you review this? I have addressed your concerns and made sure that this works on both Mac OS 10.11 and 10.12.

I am not sure how to fix circleci as it is complaining about Xcode not having its license agreement accepted on the CI.

hjanuschka commented on fastlane_core/lib/fastlane_core/device_manager.rb:
```diff
@@ -194,19 +194,45 @@ def reset_all
+      def copy_logs(device, log_identity, dest)
+        dest = File.expand_path(dest)
+        os_version = ''
+        Open3.popen3("sw_vers -productVersion") do |stdin, stdout, stderr, wait_thr|
```
can you please change it to use `FastlaneCore::CommandExecutor` ? to stay in line with other command calls?

hjanuschka commented on fastlane_core/lib/fastlane_core/device_manager.rb:
```diff
@@ -194,19 +194,45 @@ def reset_all
-        FileUtils.cp_r(logarchive_src, dest)
-        UI.success "Copying file '#{logarchive_src}' to '#{dest}'..."
+        FileUtils.mkdir_p(logarchive_dst)
+        FileUtils.cp_r("#{logarchive_src}/.", logarchive_dst)
```
🤔 that was the deal breaker?

lyndsey-ferguson replied:
This copies the _contents_ into the logarchive bundle which is a directory. Otherwise we are copying the src directory `blah.logarchive` into another directory named `specialblah.logarchive` => `blah.logarchive/specialblah.logarchive`

lyndsey-ferguson replied:
Swap specialblah with special. I'm typing on an iPhone

lyndsey-ferguson replied:
Err specialblah with blah

hjanuschka reviewed (changes requested):
@lyndsey-ferguson one comment left, i think its almost done 👍  really big thanks for your contribution. tested it locally archive is "openable"

hjanuschka commented:
about CI  - can you push a empty commit to rerun circle? - other PR's seem to work now.

lyndsey-ferguson commented:
Once I am back at a computer, I'll update the code and the subsequent push will rerun circle and all will be well :smile:

lyndsey-ferguson commented:
@hjanuschka can you look at this now?

hjanuschka commented on scan/spec/test_command_generator_spec.rb:
```diff
@@ -59,9 +59,11 @@
-    allow(Open3).to receive(:popen3).with("xcrun simctl list devices").and_yield(nil, response, nil, nil)
+    list_device_response = "list_device response"
+    allow(list_device_response).to receive(:read).and_return(@valid_simulators)
+    allow(Open3).to receive(:popen3).with("xcrun simctl list devices").and_yield(nil, list_device_response, nil, nil)
```
should'nt that be a commandexecutor?

hjanuschka replied:
ahhh i see - this was in before

lyndsey-ferguson replied:
Hmm..I'd be changing Open3's all over the place, how about I revert the Open3 change here?

hjanuschka reviewed (changes requested):
see my note about the spec

lyndsey-ferguson commented:
@hjanuschka I addressed your comment to reverting that change to be the way it was. `device_manager.rb` uses a `Open3` frequently and it seems prudent to make that change in a separate PR. I use Jira in-house for tracking work, do you recommend making a new Issue here in Github to change the use of `Open3` to `FastlaneCore::CommandExecutor.execute` ?

hjanuschka commented:
PR is good, lets wait for @DanToml  to confirm it fixes the issues on circle ;)
@lyndsey-ferguson yeah - lets create a issue, and discuss the future of open3 👍  - i am not sure if there is a reason they are in.....

lyndsey-ferguson commented:
Issue here: https://github.com/fastlane/fastlane/issues/8140

lacostej commented on fastlane_core/lib/fastlane_core/device_manager.rb:
```diff
@@ -194,19 +194,43 @@ def reset_all
       end
 
-      def copy_logarchive(device, dest)
+      def copy_logs(device, log_identity, dest)
```
are those 2 new methods meant to be public or could they be `private` ?

lacostej replied:
And maybe add some documentation. Is dest meant to be a directory`? Etc.

lacostej commented on fastlane_core/lib/fastlane_core/device_manager.rb:
```diff
@@ -194,19 +194,43 @@ def reset_all
+
+        logfile_dst = File.join(dest, "system-#{log_identity}.log")
+
+        FileUtils.remove_file(logfile_dst, force: true)
```
We may want to use `rm_f` here? It's even restricted to files as well.

lacostej commented:
Thanks for the PR! I added a couple of cleanups.

lacostej commented on fastlane_core/lib/fastlane_core/device_manager.rb:
```diff
@@ -234,6 +219,47 @@ def launch(device)
+
+        logfile_dst = File.join(logs_destination_dir, "system-#{log_identity}.log")
+
+        FileUtils.rm_rf(logfile_dst)
```
that one should be `rm_f` :)

lacostej replied:
and missing the  `FileUtils.mkdir_p(logarchive_dst)`:)

lyndsey-ferguson replied:
`FileUtils.mkdir_p` is not necessary. It _was_ necessary in the `copy_logarchive` because an actual directory is being made

lyndsey-ferguson replied:
Unless... I guess we should make the directory if it doesn't exist

lyndsey-ferguson replied:
Probably accounts for the error that @DanToml found

lacostej reviewed (approved):
Please make the last small change! Thanks

lyndsey-ferguson commented:
@lacostej your comments have been addressed

hjanuschka commented:
🚀
